File: apps/web/src/api/operations.ts
import { apiRequest } from "./client";

export interface PersonRef {
  0: string;
  1: string;
}

export interface ReminderCard {
  id: string;
  effective_at: string;
  revision: number;
}

export type PlannerStatus =
  | "not_required"
  | "needs_transfer"
  | "transferred"
  | "update_required"
  | "no_longer_relevant";

export interface WorkItemCardData {
  id: string;
  type: string;
  status: string;
  title: string;
  description: string | null;
  priority: string;
  planner_status: PlannerStatus;
  topic_id: string | null;
  topic_name: string | null;
  people: PersonRef[];
  due_at: string | null;
  next_follow_up_at: string | null;
  waiting_since: string | null;
  completed_at: string | null;
  updated_at: string;
  effective_at: string | null;
  overdue: boolean;
  revision: number;
  reminder: ReminderCard | null;
}

export interface PageResponse<T> {
  items: T[];
  limit: number;
  offset: number;
  has_more: boolean;
  timezone?: string;
}

export interface ActivityEntry {
  id: string;
  work_item_id: string;
  title: string;
  event_type: string;
  created_at: string;
}

export interface TodayOverviewResponse {
  timezone: string;
  summary: {
    overdue: number;
    due_today: number;
    follow_ups: number;
    waiting_overdue: number;
    questions: number;
    inbox: number;
    planner_queue: number;
  };
  focus: WorkItemCardData[];
  later_today: {
    items: WorkItemCardData[];
    has_more: boolean;
  };
}

export interface OverviewWorkItem {
  item: WorkItemCardData;
  needs_inbox: boolean;
}

export interface OverviewInboxItem {
  id: string;
  kind: "draft" | "work_item" | "note";
  title: string;
  excerpt: string;
  status: string;
  reasons: string[];
  occurred_at: string;
  item_count: number;
}

export interface OverviewColumn<T> {
  items: T[];
  total: number;
  has_more: boolean;
}

export interface OverviewResponse {
  timezone: string;
  today: OverviewColumn<OverviewWorkItem>;
  tomorrow: OverviewColumn<OverviewWorkItem>;
  inbox: OverviewColumn<OverviewInboxItem>;
}

export interface TopicSummary {
  id: string;
  name: string;
  description: string | null;
  open_count: number;
  overdue_count: number;
  follow_up_count: number;
  waiting_count: number;
  next_deadline: string | null;
}

export interface PersonSummary {
  id: string;
  display_name: string;
  role: string | null;
  open_item_count: number;
  follow_up_count: number;
  waiting_count: number;
  question_count: number;
  last_activity: string;
}

export type PeopleScope = "work" | "recent" | "all";

export interface NamedEntry {
  id: string;
  name: string;
  subtitle: string | null;
}

export interface NoteEntry {
  id: string;
  content: string;
  created_at: string;
}

export interface AgendaEntry {
  group_kind: "person" | "topic" | "unassigned";
  group_id: string | null;
  group_label: string;
  item: WorkItemCardData;
}

export type WorkItemAction =
  | "complete"
  | "reopen"
  | "cancel"
  | "reschedule"
  | "reschedule_preset"
  | "reschedule_text"
  | "snooze"
  | "add_note"
  | "waiting_received"
  | "agenda_discussed"
  | "question_answered"
  | "defer"
  | "convert_to_task"
  | "add_result"
  | "add_decision"
  | "archive"
  | "edit"
  | "planner_transferred"
  | "planner_not_required"
  | "planner_update_required"
  | "planner_needs_transfer";

export type ReschedulePreset =
  "later_today" | "tomorrow_morning" | "next_working_day" | "next_week";

export type RescheduleSelection =
  | {
      action: "reschedule";
      local_date: string;
      local_time: string;
    }
  | {
      action: "reschedule_preset";
      preset: ReschedulePreset;
    }
  | {
      action: "reschedule_text";
      phrase: string;
    };

export interface ActionPayload {
  action: WorkItemAction;
  client_action_id: string;
  expected_revision: number;
  content?: string;
  preset?: ReschedulePreset;
  phrase?: string;
  local_date?: string;
  local_time?: string;
  duration_minutes?: number;
  reminder_id?: string;
  reminder_revision?: number;
  title?: string;
  description?: string | null;
  item_type?: string;
  priority?: string;
  topic_id?: string | null;
  person_ids?: string[];
  date_changed?: boolean;
}

export interface ActionResponse {
  changed: boolean;
  work_item?: WorkItemCardData;
  reminder_id?: string;
  decision_id?: string;
}

export const operationsKeys = {
  all: ["operations"] as const,
  overview: ["operations", "overview"] as const,
  todayOverview: ["operations", "today", "overview"] as const,
  tomorrow: ["operations", "tomorrow"] as const,
};

function query(path: string, values: Record<string, string | number | undefined>) {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(values)) {
    if (value !== undefined && value !== "") params.set(key, String(value));
  }
  return `${path}?${params.toString()}`;
}

export const getTodayOverview = () =>
  apiRequest<TodayOverviewResponse>("/api/v1/today/overview");

export const getOverview = () => apiRequest<OverviewResponse>("/api/v1/overview");

export const getToday = (section: string, offset = 0) =>
  apiRequest<PageResponse<WorkItemCardData>>(
    query("/api/v1/today", { section, limit: 20, offset }),
  );

export const getTomorrow = (offset = 0) =>
  apiRequest<PageResponse<WorkItemCardData>>(
    query("/api/v1/tomorrow", { limit: 20, offset }),
  );

export const getTopics = (q: string, offset: number) =>
  apiRequest<PageResponse<TopicSummary>>(query("/api/v1/topics", { q, limit: 20, offset }));

export const getTopic = (id: string) =>
  apiRequest<{ id: string; name: string; description: string | null }>(
    `/api/v1/topics/${id}`,
  );

export const getTopicContent = <T>(id: string, section: string, offset: number) =>
  apiRequest<PageResponse<T>>(
    query(`/api/v1/topics/${id}/content`, { section, limit: 20, offset }),
  );

export const getPeople = (q: string, offset: number, scope: PeopleScope = "work") =>
  apiRequest<PageResponse<PersonSummary>>(
    query("/api/v1/people", { q, scope, limit: 20, offset }),
  );

export const getPerson = (id: string) =>
  apiRequest<{
    id: string;
    display_name: string;
    role: string | null;
    notes: string | null;
  }>(`/api/v1/people/${id}`);

export const getPersonContent = <T>(id: string, section: string, offset: number) =>
  apiRequest<PageResponse<T>>(
    query(`/api/v1/people/${id}/content`, { section, limit: 20, offset }),
  );

export const getAgenda = (groupKind: string, offset: number) =>
  apiRequest<PageResponse<AgendaEntry>>(
    query("/api/v1/agenda", {
      group_kind: groupKind || undefined,
      limit: 40,
      offset,
    }),
  );

export const runWorkItemAction = (id: string, payload: ActionPayload) =>
  apiRequest<ActionResponse>(`/api/v1/work-items/${id}/actions`, {
    method: "POST",
    body: JSON.stringify(payload),
  });

File: src/flowmate/api/routes/operations.py
from datetime import UTC, date, datetime, time, timedelta
from typing import Annotated, Literal, cast
from uuid import UUID
from zoneinfo import ZoneInfo

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.ext.asyncio import AsyncSession

from flowmate.api.dependencies import get_session
from flowmate.auth.dependencies import PwaIdentity, require_csrf, require_pwa_session
from flowmate.core.config import Settings, get_settings
from flowmate.reminders.actions import snooze_work_item_reminder
from flowmate.reminders.preferences import (
    EffectiveNotificationPreferences,
    NotificationDefaults,
    get_effective_notification_preferences,
)
from flowmate.reminders.timezone import resolve_local_datetime
from flowmate.task_engine.enums import PlannerStatus, WorkItemPriority, WorkItemType
from flowmate.task_engine.management import (
    InvalidWorkItemTransitionError,
    StaleWorkItemError,
    add_decision_from_work_item,
    add_work_item_note,
    archive_work_item,
    bind_client_action,
    cancel_work_item,
    change_planner_status,
    complete_work_item,
    convert_work_item_to_task,
    edit_work_item,
    mark_waiting_received,
    reopen_work_item,
    reschedule_work_item,
)
from flowmate.task_engine.operational import (
    build_work_item_cards,
    dashboard_snapshot,
    get_owned_person,
    get_owned_topic,
    list_agenda,
    list_context_content,
    list_people_summary,
    list_today_section,
    list_topics_summary,
)
from flowmate.task_engine.queries import PersonScope

router = APIRouter(prefix="/api/v1", tags=["pwa-operations"])

TodaySection = Literal["overdue", "due_today", "follow_ups", "waiting", "questions"]
TopicSection = Literal["active", "people", "notes", "decisions", "history"]
PersonSection = Literal[
    "follow_ups", "waiting", "questions", "topics", "notes", "history"
]


class WorkItemActionBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    client_action_id: UUID
    expected_revision: int = Field(ge=0)


class SimpleWorkItemAction(WorkItemActionBase):
    action: Literal[
        "complete",
        "reopen",
        "cancel",
        "waiting_received",
        "agenda_discussed",
        "question_answered",
        "convert_to_task",
        "archive",
        "planner_transferred",
        "planner_not_required",
        "planner_update_required",
        "planner_needs_transfer",
    ]


class ContentWorkItemAction(WorkItemActionBase):
    action: Literal["add_note", "add_result", "add_decision"]
    content: str = Field(min_length=1, max_length=10_000)


class DateWorkItemAction(WorkItemActionBase):
    action: Literal["reschedule", "defer"]
    local_date: date
    local_time: time


class SnoozeWorkItemAction(WorkItemActionBase):
    action: Literal["snooze"]
    duration_minutes: int | None = Field(default=None, ge=1, le=10_080)
    local_date: date | None = None
    local_time: time | None = None
    reminder_id: UUID
    reminder_revision: int = Field(ge=0)


class EditWorkItemAction(WorkItemActionBase):
    action: Literal["edit"]
    title: str = Field(min_length=1, max_length=10_000)
    description: str | None = Field(default=None, max_length=20_000)
    item_type: WorkItemType
    priority: WorkItemPriority
    topic_id: UUID | None = None
    person_ids: list[UUID] = Field(default_factory=list, max_length=50)
    date_changed: bool = False
    local_date: date | None = None
    local_time: time | None = None


WorkItemActionRequest = Annotated[
    SimpleWorkItemAction
    | ContentWorkItemAction
    | DateWorkItemAction
    | SnoozeWorkItemAction
    | EditWorkItemAction,
    Field(discriminator="action"),
]


async def _preferences(
    session: AsyncSession, identity: PwaIdentity, settings: Settings
) -> EffectiveNotificationPreferences:
    return await get_effective_notification_preferences(
        session,
        identity.user.id,
        NotificationDefaults.from_settings(settings),
    )


def _clock() -> datetime:
    return datetime.now(UTC)


def _page_payload(page: object, *, timezone: str | None = None) -> dict[str, object]:
    payload = {
        "items": page.items,  # type: ignore[attr-defined]
        "limit": page.limit,  # type: ignore[attr-defined]
        "offset": page.offset,  # type: ignore[attr-defined]
        "has_more": page.has_more,  # type: ignore[attr-defined]
    }
    if timezone is not None:
        payload["timezone"] = timezone
    return payload


@router.get("/dashboard")
async def dashboard(
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> dict[str, object]:
    preferences = await _preferences(session, identity, settings)
    payload = await dashboard_snapshot(
        session, identity.user.id, now=_clock(), preferences=preferences
    )
    return {"timezone": preferences.timezone, **payload}


@router.get("/today")
async def today(
    section: TodaySection,
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    settings: Annotated[Settings, Depends(get_settings)],
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, object]:
    preferences = await _preferences(session, identity, settings)
    page = await list_today_section(
        session,
        identity.user.id,
        section,
        now=_clock(),
        preferences=preferences,
        limit=limit,
        offset=offset,
    )
    return {"section": section, **_page_payload(page, timezone=preferences.timezone)}


@router.get("/topics")
async def topics(
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    q: str | None = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, object]:
    return _page_payload(
        await list_topics_summary(
            session,
            identity.user.id,
            query=q,
            now=_clock(),
            limit=limit,
            offset=offset,
        )
    )


@router.get("/topics/{topic_id}")
async def topic_details(
    topic_id: UUID,
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
) -> dict[str, object]:
    topic = await get_owned_topic(session, identity.user.id, topic_id)
    if topic is None:
        raise HTTPException(status_code=404, detail="Topic not found")
    return {
        "id": topic.id,
        "name": topic.name,
        "description": topic.description,
        "created_at": topic.created_at,
        "updated_at": topic.updated_at,
    }


@router.get("/topics/{topic_id}/content")
async def topic_content(
    topic_id: UUID,
    section: TopicSection,
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, object]:
    if await get_owned_topic(session, identity.user.id, topic_id) is None:
        raise HTTPException(status_code=404, detail="Topic not found")
    page = await list_context_content(
        session,
        identity.user.id,
        owner_type="topic",
        owner_id=topic_id,
        section=section,
        now=_clock(),
        limit=limit,
        offset=offset,
    )
    return {"section": section, **_page_payload(page)}


@router.get("/people")
async def people(
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    q: str | None = None,
    scope: PersonScope = "work",
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, object]:
    return _page_payload(
        await list_people_summary(
            session,
            identity.user.id,
            scope=scope,
            query=q,
            now=_clock(),
            limit=limit,
            offset=offset,
        )
    )


@router.get("/people/{person_id}")
async def person_details(
    person_id: UUID,
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
) -> dict[str, object]:
    person = await get_owned_person(session, identity.user.id, person_id)
    if person is None:
        raise HTTPException(status_code=404, detail="Person not found")
    return {
        "id": person.id,
        "display_name": person.display_name,
        "role": person.role,
        "notes": person.notes,
        "created_at": person.created_at,
        "updated_at": person.updated_at,
    }


@router.get("/people/{person_id}/content")
async def person_content(
    person_id: UUID,
    section: PersonSection,
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, object]:
    if await get_owned_person(session, identity.user.id, person_id) is None:
        raise HTTPException(status_code=404, detail="Person not found")
    page = await list_context_content(
        session,
        identity.user.id,
        owner_type="person",
        owner_id=person_id,
        section=section,
        now=_clock(),
        limit=limit,
        offset=offset,
    )
    return {"section": section, **_page_payload(page)}


@router.get("/agenda")
async def agenda(
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_pwa_session)],
    settings: Annotated[Settings, Depends(get_settings)],
    group_kind: Literal["person", "topic", "unassigned"] | None = None,
    group_id: UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 40,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> dict[str, object]:
    preferences = await _preferences(session, identity, settings)
    page = await list_agenda(
        session,
        identity.user.id,
        group_kind=group_kind,
        group_id=group_id,
        now=_clock(),
        limit=limit,
        offset=offset,
    )
    return _page_payload(page, timezone=preferences.timezone)


def _target_datetime(
    payload: DateWorkItemAction | SnoozeWorkItemAction, timezone: str
) -> datetime:
    if payload.local_date is None or payload.local_time is None:
        raise HTTPException(status_code=422, detail="Local date and time are required")
    return resolve_local_datetime(
        payload.local_date, payload.local_time.replace(tzinfo=None), ZoneInfo(timezone)
    ).astimezone(UTC)


@router.post("/work-items/{work_item_id}/actions")
async def work_item_action(
    work_item_id: UUID,
    payload: WorkItemActionRequest,
    session: Annotated[AsyncSession, Depends(get_session)],
    identity: Annotated[PwaIdentity, Depends(require_csrf)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> dict[str, object]:
    bind_client_action(session, payload.client_action_id)
    user_id = identity.user.id
    try:
        if payload.action in {"complete", "agenda_discussed", "question_answered"}:
            result = await complete_work_item(
                session,
                user_id,
                work_item_id,
                None,
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "reopen":
            result = await reopen_work_item(
                session,
                user_id,
                work_item_id,
                None,
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "cancel":
            result = await cancel_work_item(
                session,
                user_id,
                work_item_id,
                None,
                expected_revision=payload.expected_revision,
            )
        elif payload.action in {"reschedule", "defer"}:
            date_payload = cast(DateWorkItemAction, payload)
            preferences = await _preferences(session, identity, settings)
            result = await reschedule_work_item(
                session,
                user_id,
                work_item_id,
                None,
                _target_datetime(date_payload, preferences.timezone),
                expected_revision=payload.expected_revision,
            )
        elif payload.action in {"add_note", "add_result"}:
            content_payload = cast(ContentWorkItemAction, payload)
            result, _ = await add_work_item_note(
                session,
                user_id,
                work_item_id,
                None,
                content_payload.content,
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "waiting_received":
            result = await mark_waiting_received(
                session,
                user_id,
                work_item_id,
                None,
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "convert_to_task":
            result = await convert_work_item_to_task(
                session,
                user_id,
                work_item_id,
                None,
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "archive":
            result = await archive_work_item(
                session,
                user_id,
                work_item_id,
                None,
                expected_revision=payload.expected_revision,
            )
        elif payload.action.startswith("planner_"):
            planner_targets = {
                "planner_transferred": PlannerStatus.TRANSFERRED,
                "planner_not_required": PlannerStatus.NOT_REQUIRED,
                "planner_update_required": PlannerStatus.UPDATE_REQUIRED,
                "planner_needs_transfer": PlannerStatus.NEEDS_TRANSFER,
            }
            result = await change_planner_status(
                session,
                user_id,
                work_item_id,
                planner_targets[payload.action],
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "edit":
            scheduled_at = None
            if payload.date_changed:
                if (payload.local_date is None) != (payload.local_time is None):
                    raise HTTPException(
                        status_code=422,
                        detail="Local date and time must be provided together",
                    )
                if payload.local_date is not None and payload.local_time is not None:
                    preferences = await _preferences(session, identity, settings)
                    scheduled_at = resolve_local_datetime(
                        payload.local_date,
                        payload.local_time.replace(tzinfo=None),
                        ZoneInfo(preferences.timezone),
                    ).astimezone(UTC)
            result = await edit_work_item(
                session,
                user_id,
                work_item_id,
                title=payload.title,
                description=payload.description,
                item_type=payload.item_type,
                priority=payload.priority,
                topic_id=payload.topic_id,
                person_ids=tuple(payload.person_ids),
                update_schedule=payload.date_changed,
                scheduled_at=scheduled_at,
                expected_revision=payload.expected_revision,
            )
        elif payload.action == "add_decision":
            result, decision = await add_decision_from_work_item(
                session,
                user_id,
                work_item_id,
                None,
                payload.content,
                expected_revision=payload.expected_revision,
            )
            await session.refresh(result.work_item)
            cards = await build_work_item_cards(
                session, user_id, [result.work_item], now=_clock()
            )
            return {
                "changed": result.changed,
                "work_item": cards[0],
                "decision_id": decision.id,
            }
        else:
            snooze_payload = cast(SnoozeWorkItemAction, payload)
            preferences = await _preferences(session, identity, settings)
            if snooze_payload.duration_minutes is not None:
                reminder, changed = await snooze_work_item_reminder(
                    session,
                    user_id,
                    snooze_payload.reminder_id,
                    None,
                    duration=timedelta(minutes=snooze_payload.duration_minutes),
                    expected_revision=snooze_payload.reminder_revision,
                )
            else:
                reminder, changed = await snooze_work_item_reminder(
                    session,
                    user_id,
                    snooze_payload.reminder_id,
                    None,
                    until=_target_datetime(snooze_payload, preferences.timezone),
                    expected_revision=snooze_payload.reminder_revision,
                )
            return {"changed": changed, "reminder_id": reminder.id}
    except StaleWorkItemError as error:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="Work item changed"
        ) from error
    except InvalidWorkItemTransitionError as error:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail=str(error)
        ) from error
    except ValueError as error:
        message = str(error)
        if "not found" in message:
            raise HTTPException(
                status_code=404, detail="Work item not found"
            ) from error
        raise HTTPException(status_code=422, detail=message) from error
    await session.refresh(result.work_item)
    cards = await build_work_item_cards(
        session, user_id, [result.work_item], now=_clock()
    )
    return {"changed": result.changed, "work_item": cards[0]}
